halo2_proofs/src/plonk/permutation/keygen.rs

```diff
@@ -162,50 +171,68 @@ impl Assembly {
-            for _ in 0..p.columns.len() {
-                let mut omega_powers = omega_powers.clone();
-                for o in &mut omega_powers {
-                    *o *= &cur;
+            parallelize(&mut deltaomega, |o, start| {
+                let mut cur = C::Scalar::DELTA.pow_vartime(&[start as u64]);
+                for omega_powers in o.iter_mut() {
+                    for v in omega_powers {
+                        *v *= &cur;
+                    }
+                    cur *= &C::Scalar::DELTA;
                 }
-
-                deltaomega.push(omega_powers);
-
-                cur *= &C::Scalar::DELTA;
-            }
+            });
         }
 
         // Compute permutation polynomials, convert to coset form.
-        let mut permutations = vec![];
-        let mut polys = vec![];
-        let mut cosets = vec![];
-        for i in 0..p.columns.len() {
-            // Computes the permutation polynomial based on the permutation
-            // description in the assembly.
-            let mut permutation_poly = domain.empty_lagrange();
-            for (j, p) in permutation_poly.iter_mut().enumerate() {
-                let (permuted_i, permuted_j) = self.mapping[i][j];
-                *p = deltaomega[permuted_i][permuted_j];
-            }
+        let mut permutations = vec![domain.empty_lagrange(); p.columns.len()];
+        {
+            parallelize(&mut permutations, |o, start| {
+                for (x, permutation_poly) in o.iter_mut().enumerate() {
+                    let i = start + x;
+                    for (j, p) in permutation_poly.iter_mut().enumerate() {
+                        let (permuted_i, permuted_j) = self.mapping[i][j];
+                        *p = deltaomega[permuted_i][permuted_j];
+                    }
+                }
+            });
+        }
 
-            // Store permutation polynomial and precompute its coset evaluation
-            permutations.push(permutation_poly.clone());
-            let poly = domain.lagrange_to_coeff(permutation_poly);
-            polys.push(poly.clone());
-            cosets.push(domain.coeff_to_extended(poly));
+        let mut polys = vec![domain.empty_coeff(); p.columns.len()];
+        {
+            parallelize(&mut polys, |o, start| {
+                for (x, poly) in o.iter_mut().enumerate() {
+                    let i = start + x;
+                    let permutation_poly = permutations[i].clone();
+                    *poly = domain.lagrange_to_coeff(permutation_poly);
+                }
+            });
         }
+
+        let mut cosets = vec![domain.empty_extended(); p.columns.len()];
+        {
+            parallelize(&mut cosets, |o, start| {
+                for (x, coset) in o.iter_mut().enumerate() {
+                    let i = start + x;
+                    let poly = polys[i].clone();
+                    *coset = domain.coeff_to_extended(poly);
+                }
+            });
+        }
+
```

Comment:
Internally `lagrange_to_coeff` and `coeff_to_extended` are already parallelly, so I'm thinking perhaps we don't need to parallelize the outer loop again.